rfc/006-semantic-versioning.md

```diff
@@ -0,0 +1,84 @@
+
+### Modes of Interprocess Communication
+
+Tendermint has two primary mediums with which it can communicate to other processes: RPC and P2P. The division marks the boundary between the internal and external components of the network. The P2P layer is used in all cases that nodes (validator, full, seed or light) need to communicate with one another whereas the RPC is for any outside process that wants to communicate with a node. The assumption made here is that all communication via the RPC is to a trusted source and thus the ability to interpret information is more important than being able to verify the information. P2P is therefore primary medium for verification. As an example, an in-browser light client would consist of verifying headers (and perhaps application state) via the p2p layer and passing the information on to the client via the RPC layer (or potentially directly via an API).
+
+The exception to this is the IBC module and relayers which are external but require verifiable data. Breaking changes to the light client verification path mean that all neighbouring chains that are connected will no longer be able to verify state transitions and thus pass messages back and forward.
+
+## Proposal
+
+Tendermint version labels will follow the syntax of [Semantic Versions 2.0.0](https://semver.org/) with a major, minor and patch version. The version components will be interpreted according to these rules:
+
+For the entire cycle of a **major version** in Tendermint:
+
+- All blocks and state data in a blockchain can be queried. All headers can be verified even across minor version changes. Nodes can both block sync and state sync from genesis to the head of the chain.
+- Nodes in a network are able to communicate and perform BFT state machine replication so long as the agreed network version is the lowest of all nodes in a network. For example, nodes using version 1.5.x and 1.2.x can operate together so long as the network version is 1.2 or lower (but still within the 1.x range). This rule essentially captures the concept of network backwards compatibility.
+- Node RPC endpoints will remain compatible with existing external clients:
+    - New endpoints may be added, but old endpoints may not be removed.
+    - Old endpoints may be extended to add new request and response fields, but requests not using those fields must function as before the change.
+- Migrations should be automatic. Upgrading of one node can happen asynchronously with respect to other nodes (although agreement of a network-wide upgrade must still occur synchronously via consensus).
+
+For the entire cycle of a **minor version** in Tendermint:
+
+- Public Go API's, for example in `node` or `abci` packages will not change in a way that requires any consumer (not just application developers) to modify their code.
```

Comment:
The mild difficulty that I see this presenting is that Go's tooling will consider two versions 'compatible' within the same major version. Using the tooling to automatically grab the latest will now grab a version of the software that breaks compile. I don't have a huge objection to this, but it is a bit out of line with the way most go projects work.

Reply:
Agreed, but note below:
> These guarantees will come into effect at release 1.0.

so I think we are safe. Possibly we should highlight that line at the top of the doc, though.